### PR TITLE
feat: adds azure identity capability

### PR DESCRIPTION
**Pull request by fardarter**
This PR adds the ability to use an Azure Managed Identity for authentication. 

- Adds optional configuration for using an Azure managed identity via new parameters: `auth_type`, `clientId`
- Adds support for the use of `DefaultAzureCredential` from the '@azure/identity' package
- Supports both system and user assigned managed identities via optionally passing `clientId` to `DefaultAzureCredential`. 

Addresses issue: https://github.com/jakeFeldman/strapi-provider-upload-azure-storage/issues/71

Can be tested by:
a) using a local authorised logged in user
b) deploying a working Strapi instance to an app service and using the system assigned identity permissions
c) assigning a working app service a user assigned managed identity and passing the identity id to the `clientId` configuration parameter.

RBAC role of Storage Blob Data Contributor over either the target account or the target container should be sufficient for upload.

### CODE REVIEW

**Comment by jakeFeldman** [comment]
Hello @fardarter thank you for the PR. Do you mind adding a description to the PR. Thanks

**Comment by fardarter** [comment]
> Hello @fardarter thank you for the PR. Do you mind adding a description to the PR. Thanks

Is this for release notes? Do you need a specific approach/format?

**Comment by fardarter** [comment]
> > Hello @fardarter thank you for the PR. Do you mind adding a description to the PR. Thanks
> 
> Is this for release notes? Do you need a specific approach/format?

@jakeFeldman Just following up here. I'm keen to get this in so happy to adjust to your needs.

**Comment by jakeFeldman** [comment]
@fardarter Thanks for the follow up. Yes, having a detailed description with the changes, how to test, etc.. would be good for the release notes and posterity.

File: .eslintrc.js [hunk 1]
```diff
@@ -27,6 +27,7 @@ module.exports = {
         'simple-import-sort/exports': 'error',
         'no-unused-vars': 'off',
         'unused-imports/no-unused-imports': 'error',
+        'no-case-declarations': 'off',
```
Comment (jakeFeldman) [inline .eslintrc.js:30]: Lets not turn this off. We can fix this by use braces to ensure proper scope in the `case`

File: src/index.ts [hunk 2]
```diff
@@ -43,18 +59,31 @@ function getFileName(path: string, file: StrapiFile) {
...
+        case 'msi':
+            const clientId = trimParam(config.clientId);
+            if (clientId != null && clientId != '') {
+                return new BlobServiceClient(
+                    serviceBaseURL,
+                    new DefaultAzureCredential({ managedIdentityClientId: clientId })
+                );
+            }
+            return new BlobServiceClient(serviceBaseURL, new DefaultAzureCredential());
     }
```
Comment (jakeFeldman) [inline src/index.ts:86]: for example:

```ts
switch (config.auth_type) {
        case 'default': {
            const account = trimParam(config.account);
            const accountKey = trimParam(config.accountKey);
            const sasToken = trimParam(config.sasToken);
            if (sasToken != '') {
                const anonymousCredential = new AnonymousCredential();
                return new BlobServiceClient(`${serviceBaseURL}${sasToken}`, anonymousCredential);
            }
            const sharedKeyCredential = new StorageSharedKeyCredential(account, accountKey);
            const pipeline = newPipeline(sharedKeyCredential);
            return new BlobServiceClient(serviceBaseURL, pipeline);
        }
        case 'msi': {
            const clientId = trimParam(config.clientId);
            if (clientId != null && clientId != '') {
                return new BlobServiceClient(
                    serviceBaseURL,
                    new DefaultAzureCredential({ managedIdentityClientId: clientId })
                );
            }
            return new BlobServiceClient(serviceBaseURL, new DefaultAzureCredential());
        }
    }
```

**Comment by fardarter** [comment]
> @fardarter Thanks for the follow up. Yes, having a detailed description with the changes, how to test, etc.. would be good for the release notes and posterity.

Have updated and will follow up on the other. I'm testing at the moment with a patch-package version of a `dist` I'm building locally with this code in all the scenarios listed. I can let you know if I run into any bugs today but I don't anticipate any.

File: .eslintrc.js (hunk 1 above)
Comment (fardarter) [reply .eslintrc.js:30]: Will update

Comment (fardarter) [reply .eslintrc.js]: Have updated. Will leave resolution to you as the person confirming acceptance of the change.

**Comment by fardarter** [comment]
> > @fardarter Thanks for the follow up. Yes, having a detailed description with the changes, how to test, etc.. would be good for the release notes and posterity.
> 
> Have updated and will follow up on the other. I'm testing at the moment with a patch-package version of a `dist` I'm building locally with this code in all the scenarios listed. I can let you know if I run into any bugs today but I don't anticipate any.

@jakeFeldman Have had successful deployments yesterday with clientId and plain msi for days before that.

File: src/index.ts [hunk 3]
```diff
@@ -43,18 +59,32 @@ function getFileName(path: string, file: StrapiFile) {
...
+        case 'msi': {
+            const clientId = trimParam(config.clientId);
+            if (clientId != null && clientId != '') {
+                return new BlobServiceClient(
+                    serviceBaseURL,
+                    new DefaultAzureCredential({ managedIdentityClientId: clientId })
+                );
+            }
+            return new BlobServiceClient(serviceBaseURL, new DefaultAzureCredential());
+        }
```
Comment (jakeFeldman) [inline src/index.ts:86]: We should just use the switch/cases default vs defining a case for it

```suggestion

    switch (config.auth_type) {
        case 'msi': {
            const clientId = trimParam(config.clientId);
            if (clientId != null && clientId != '') {
                return new BlobServiceClient(
                    serviceBaseURL,
                    new DefaultAzureCredential({ managedIdentityClientId: clientId })
                );
            }
            return new BlobServiceClient(serviceBaseURL, new DefaultAzureCredential());
        }
       default: {
            const account = trimParam(config.account);
            const accountKey = trimParam(config.accountKey);
            const sasToken = trimParam(config.sasToken);
            if (sasToken != '') {
                const anonymousCredential = new AnonymousCredential();
                return new BlobServiceClient(`${serviceBaseURL}${sasToken}`, anonymousCredential);
            }
            const sharedKeyCredential = new StorageSharedKeyCredential(account, accountKey);
            const pipeline = newPipeline(sharedKeyCredential);
            return new BlobServiceClient(serviceBaseURL, pipeline);
        }
}
```

File: src/index.ts [hunk 4]
```diff
@@ -43,18 +59,32 @@ function getFileName(path: string, file: StrapiFile) {
...
-    const account = trimParam(config.account);
-    const accountKey = trimParam(config.accountKey);
-    const sasToken = trimParam(config.sasToken);
     const serviceBaseURL = getServiceBaseUrl(config);
-    // if accountKey doesn't contain value return below line
-    if (sasToken != '') {
-        const anonymousCredential = new AnonymousCredential();
-        return new BlobServiceClient(`${serviceBaseURL}${sasToken}`, anonymousCredential);
+
+    switch (config.auth_type) {
```
Comment (jakeFeldman) [inline src/index.ts:64]: all config vars are camelCase. Will you update to the var name to `authType` thanks

**Review by jakeFeldman — COMMENTED** [review]
This looks good. Just a few other clean up items and I would be happy to merge this.

File: src/index.ts (hunk 3 above)
Comment (fardarter) [reply src/index.ts:86]: The explicit cases are what are doing the type narrowing on the tags, so moving them to the default case means TS doesn't recognise the params as belonging to that specific branch of the union. 

I can change the name from `default` to `accountKeyOrSas` if you find it more informative?

File: src/index.ts (hunk 4 above)
Comment (fardarter) [reply src/index.ts:64]: Right, will change

Comment (fardarter) [reply src/index.ts]: Amended

File: src/index.ts (hunk 3 above)
Comment (fardarter) [reply src/index.ts]: Have added a never case to the default so there isn't a fall-through case. 

Not sure where the type checking happens on the strapi side, if at all?

Comment (jakeFeldman) [reply src/index.ts]: That makes sense. Thanks for adding the default case.

File: src/index.ts (hunk 4 above)
Comment (jakeFeldman) [reply src/index.ts]: Thank you.

**Review by jakeFeldman — APPROVED** [review]
Thanks for the PR. This looks good. Let me know, if you're interested in helping maintain the plugin. I would appreciate the help.

**Comment by fardarter** [comment]
> Thanks for the PR. This looks good. Let me know, if you're interested in helping maintain the plugin. I would appreciate the help.

Thanks for the merge. Looking forward to seeing it in a new release.

Honestly don't know what I'd be signing up for. 

I'm also maybe not the best candidate as I'm mainly an ops person. I stop interfacing with Strapi once it's deployed and I can use identities to access the things I need (trying to run secretless infra). I don't know much about the actual plugin interop.

**Comment by jakeFeldman** [comment]
Hello, I just published version 3.4.0-beta. You can test it by adding the beta tag to your install script. 

```sh
yarn add strapi-provider-upload-azure-storage@beta
```

If you want to test the version and make sure that both authTypes are working I can publish v3.4.0.

No worries wrt to maintenance. The amount of work is low/minimal, I just don't use Azure much these days.

**Comment by jakeFeldman** [comment]
Hey @fardarter, I just published version `v3.4.0-beta`. You can test it by adding the beta tag to your install script. 

```sh
yarn add strapi-provider-upload-azure-storage@beta
```

If you want to test the version and make sure that both authTypes are working I can publish v3.4.0.

No worries wrt to maintenance. The amount of work is low/minimal, I just don't use Azure much these days.

**Comment by fardarter** [comment]
> Hey @fardarter, I just published version `v3.4.0-beta`. You can test it by adding the beta tag to your install script.
> 
> ```shell
> yarn add strapi-provider-upload-azure-storage@beta
> ```
> 
> If you want to test the version and make sure that both authTypes are working I can publish v3.4.0.
> 
> No worries wrt to maintenance. The amount of work is low/minimal, I just don't use Azure much these days.

Thanks, I'll do that. 

Do you mainly just need someone to triage issues?

**Comment by fardarter** [comment]
> Hey @fardarter, I just published version `v3.4.0-beta`. You can test it by adding the beta tag to your install script.
> 
> ```shell
> yarn add strapi-provider-upload-azure-storage@beta
> ```
> 
> If you want to test the version and make sure that both authTypes are working I can publish v3.4.0.
> 
> No worries wrt to maintenance. The amount of work is low/minimal, I just don't use Azure much these days.

Working with beta package for `msi` flow.

**Comment by jakeFeldman** [comment]
Great to hear. Version 3.4.0 is available now. Thanks again for the PR

File: README.md [hunk 5]
```diff
@@ -35,12 +35,12 @@ To enable the provider, create or edit the file at `./config/plugins.js`.
 This is an example `plugins.js` file for Azure storage:
 
 ```js
-
 module.exports = ({ env }) => ({
   upload: {
     config: {
       provider: "strapi-provider-upload-azure-storage",
       providerOptions: {
+        auth_type: env("STORAGE_AUTH_TYPE", "default"),
```
Comment (Achmadrendra1) [inline README.md]: why wihout this line produces error ? did msoft change something?

Comment (fardarter) [reply README.md]: @Achmadrendra1 You need to include an `authType` now. It's so to distinguish between the two key types. If you're using the original variables, all you need to do it include that line.

Comment (jakeFeldman) [reply README.md]: Sorry. In hindsight I should have published this as a major version since that was a breaking change.

**Pull request merged**